arch/arc/core/cache.c

```diff
@@ -218,8 +218,9 @@ int arch_icache_flush_and_invd_range(void *addr, size_t size)
 
 static int init_dcache(void)
 {
-
-	arch_dcache_enable();
+	if (IS_ENABLED(CONFIG_DCACHE)) {
+		arch_dcache_enable();
+	}
```

Comment:
The problem here is that the architecture must not use `arch_dcache_enable` directly, but instead they have to rely on `sys_cache_data_enable`.

`sys_cache_data_enable` collapse to a NOP when `!CONFIG_CACHE_MANAGEMENT` or `!CONFIG_DCACHE`.

Reply:
fine for me, will change it

Reply:
Forgive me if I'm wrong, but that would never enable caches unless both of these settings were enabled, and it makes sense to enable the CPU cache but not the Cache API interface. 
If using `arch_` functions is wrong in this context, perhaps the suggestion should be to use a guarded `cache_data_enable()` instead?

Reply:
> Forgive me if I'm wrong, but that would never turn on caches unless both of these settings were enabled

Correct, and that is explained in https://docs.zephyrproject.org/latest/hardware/cache/index.html indeed.

`CONFIG_DCACHE` is telling you that a `DCACHE` is present but whether that is actually used or not it is up to `CONFIG_CACHE_MANAGEMENT` option.

> it makes sense to be able to use the CPU cache but not the Cache API interface.

Does it though? Before the cache API was introduced every single architecture was doing something different and custom, the cache API was introduced exactly to put order in the cache management chaos. So the path forward here is just to move to the new cache API, not adding workarounds.

Reply:
I agree with @pillo79, CONFIG_CACHE_MANAGEMENT might be disabled manually, but still CONFIG_DCACHE enabled. In such a case I would expect not to be able to enable or disable the caches from the application, but still the data cache to be active.

    static ALWAYS_INLINE void sys_cache_data_enable(void)
    {
    #if defined(CONFIG_CACHE_MANAGEMENT) && defined(CONFIG_DCACHE)
        cache_data_enable();
    #endif
    }

I'm rather thinking towards calling cache_data_enable, with IS_ENABLED(CONFIG_DCACHE) around it.

Reply:
> `CONFIG_DCACHE` is telling you that a `DCACHE` is present but whether that is actually used or not it is up to `CONFIG_CACHE_MANAGEMENT` option.

This is not represented like this in the help for the option:

    This option enables the support for the data cache (d-cache).

There might be a difference being cache available and cache being used/activated.

Reply:
@carlocaione thanks for the link! Still confused though... quoting from there (emphasis mine):
> `CONFIG_DCACHE` / `CONFIG_ICACHE`: [...] All the code paths related to cache control must be conditionally enabled depending on these symbols. **When the symbol is set the cache is considered enabled and used**.
> These symbols say nothing about the actual API interface exposed to the user. For example a platform using the data cache can enable the `CONFIG_DCACHE` symbol and use some HAL exported function in some platform-specific code to enable and manage the d-cache.

> `CONFIG_CACHE_MANAGEMENT`: this option must be selected when the cache operations are exposed to the user through a standard API (see [Cache API](https://docs.zephyrproject.org/latest/hardware/cache/index.html#cache-api)).

Reply:
when the cache API was introduced we could not move all the cache implementation over right away, it was just too much work. So where possible we simply used `CONFIG_DCACHE` around the legacy cache operation code and we added `CONFIG_DCACHE` to be set at SoC / platform level to say: "the SoC has a functional DCACHE that is being used by some code".

For legacy code we moved from:
```
  custom_cache_op();
```
to:
```
#ifdef CONFIG_DCACHE
  custom_cache_op();
#endif
```

So, a quick recap is:
- `CONFIG_CPU_HAS_DCACHE`: the CPU has a DCACHE, it can be used or not
- `CONFIG_DCACHE`: the CPU has a DCACHE that is being used somehow (this was done to support legacy cache operations)
- `CONFIG_CACHE_MANAGEMENT`: the CPU has a DCACHE that is being used using the cache API

So, back to this PR. Is this PR correct? It is, but it is simply fixing a legacy usage that it should (and it is) deprecated.

Reply:
Thanks for the clarification, but I think one version in between is missing.

- `CONFIG_CPU_HAS_DCACHE`: the CPU has a DCACHE, it can be used or not
- `CONFIG_DCACHE`: the CPU has a DCACHE that is being used somehow (this was done to support legacy cache operations)
- `???`: the CPU as DCACHE which is active, but it is not accessible using the cache API
- `CONFIG_CACHE_MANAGEMENT`: the CPU has a DCACHE that is being used using the cache API

I will in the meantime go ahead with moving forward to sys_cache_data_enable as CONFIG_DCACHE seems to be somewhat deprecated.

Reply:
> the CPU as DCACHE which is active, but it is not accessible using the cache API

Well, that would be `CONFIG_DCACHE=y` and `CONFIG_CACHE_MANAGEMENT=n`.

That is: you have something managing the DCACHE using some kind of custom API that it is not the zephyr cache API. In this case your code should use the custom / HAL API to manage the cache and now you see why we needed a formal, shared API for that, because we needed to abstract the cache operations away from the actual implementation.

Reply:
But with `CONFIG_DCACHE=y` and `CONFIG_CACHE_MANAGEMENT=n` sys_cache_data_enable will turn into NOP. Therefore I won't be able to use the cache management API.
I totally agree with you that having an abstract cache operations API is a good thing to have. I'm just saying that having this API available doesn't necessarily mean that one wants to automatically enable the cache. These are two separate things for me.

Reply:
> But with CONFIG_DCACHE=y and CONFIG_CACHE_MANAGEMENT=n sys_cache_data_enable will turn into NOP

Yes, exactly.

You can have `CONFIG_DCACHE=y` and `CONFIG_CACHE_MANAGEMENT=n` in two cases:

1. Your SoC has `sys_cache_data_enabled` implemented but you do not want to use it, this is why you are setting `CONFIG_CACHE_MANAGEMENT=n`, so `sys_cache_data_enable()` is returning `-ENOTSUP`.
2. Your SoC doesn't have `sys_cache_data_enabled` implemented, but to enable the DCACHE it is relying on some custom API, so `sys_cache_data_enable()` is returning `-ENOTSUP` because the API that you want to use is different (custom).